# Chinese Syllable Count Generator Working Draft Notes
*Designed, constructed, and documented by Jeff Heller, Data and Project Coordinator for the Department of East Asian Studies, Princeton University.*

*With expertise and assistance from fellow Princetonians:*

*Anna Shields, Professor of East Asian Studies*

*Joshua Seufert, Chinese Studies Librarian, The East Asian Library and the Gest Collection*

*Thomas Ventimiglia, Technical Support Specialist, The East Asian Library and the Gest Collection*

This document details the process of experimentation related to the construction of the Chinese Syllable Count Generator function. The function takes romanized Chinese text of any size or word count, evaluates the number of syllables based on the supplied Latin characters, and outputs a list of integers representing the syllables of each identified word.

The initial stages of development tests three methods of storing data for optimized processing comparison: Python list, NumPy Array, and Pandas DataFrame.

In [1]:
import pandas as pd
import numpy as np

Structured systems are the basis of Chinese romanization with accepted combinations of "initials" (the initial sound of a syllable, e.g., the first consonant) and "finals" (the ending sound of a syllable, e.g., vowels that sometimes end with a consonant). For example, one syllable in pinyin can be as small as one letter and as large as six letters. Therefore, detecting a syllable can occur in one of two ways:
1. A sequence of letters compared to a list of accepted pinyin syllables. If the sequence appears in the list, or if the sequence is part of a longer accepted term ("tia" by itself is not an accepted syllable, however "tiao" and "tian" are), then an additional letter from the pinyin can be added to the sequence and reevaluated. If the sequence does not appear in the list and is not part of a longer accepted term, then the last character of the sequence becomes the first character of the next syllable, and the syllable count incremented.
2. A sequence of letters is processed incrementally with the initials identified first, and then the finals after the appearance of the first vowel. The identified initial and final values act as indexes to a table (either a NumPy Array or Pandas DataFrame) of boolean values that evaluate the initial and final pairing. If the table returns False, then the final is compared to a list of other final values to determine whether it can be part of a more extended final. If it is, then the function continues onto the next incremented letter. Otherwise, the function determines that the current letter is the start of the next syllable and increments the number of syllables for that word.

Given these possibilities, two metrics need to be considered to settle on the proper method for the evaluation function: speed and size. I hypothesize that comparing a sequence of letters to a list of terms can be faster; however, the list itself may be excessively large. Evaluating sequences by their combination of initials and finals, however, may take longer but results in a much smaller package. The following exercises test each method on a smaller scale to get a general sense of the speed vs. size issue.

## Setup for Testing

The code below imports two CSV files. The first contains the list of acceptable pinyin syllables (e.g. "yang," "wei"). The second contains an array of 1 values with a column header of initials and row headers with finals. Empty values convert to 0, and both 1 and 0 values convert to the boolean values True and False, respectively in their array and DataFrame forms.

In [51]:
import csv
import sys

## LIST ##
# Import list of all pinyin combinations
smList = []
with open("data/pinyinList.csv") as f:
    r = csv.reader(f)
    for row in r:
        smList += row

## NUMPY/DATAFRAME ##
# List of initials, finals for Numpy evaluation
PY = np.genfromtxt('data/pinyinDF.csv', delimiter=',', dtype=str)
initList = list(PY[1:, 0])
finList = list(PY[0, 1:])

# Import Numpy array from CSV
smAr = PY[1:, 1:] == '1'

# Import dataframe from CSV
smDF = pd.read_csv('data/pinyinDF.csv', header=0, index_col=0, dtype={'INDEX': str}
                   ).fillna(0).astype('bool')

# Print examples of each form of reference data.
print('Python List:')
print(str(smList[0:30]))
print('Displaying 30 of ' + str(len(smList)) +
      ' possible pinyin combinations ' + '\n')
print('——————————————————\n')
print('# Header lists for use in indexing the array/DataFrame')
print('Initials List:')
print(initList)
print('Finals List:')
print(str(finList) + '\n')
print('——————————————————\n')
print('Numpy Array:')
print(str(smAr[0]))
print('Displaying 1 of ' + str(len(smAr)) + ' rows within the array' + '\n')
print('Pandas DataFrame:')
print(smDF.head(3))

Python List:
['a', 'ai', 'ao', 'an', 'ang', 'e', 'ei', 'en', 'eng', 'er', 'o', 'ou', 'yi', 'ya', 'yao', 'ye', 'you', 'yan', 'yang', 'yin', 'ying', 'yong', 'wu', 'wa', 'wo', 'wei', 'wai', 'wan', 'wen', 'wang']
Displaying 30 of 407 possible pinyin combinations 

——————————————————

# Header lists for use in indexing the array/DataFrame
Initials List:
['ø', 'b', 'p', 'm', 'f', 'd', 't', 'n', 'l', 'z', 'c', 's', 'zh', 'ch', 'sh', 'r', 'j', 'q', 'x', 'g', 'k', 'h', 'y', 'w']
Finals List:
['a', 'ai', 'ao', 'an', 'ang', 'e', 'ei', 'en', 'eng', 'er', 'o', 'ong', 'ou', 'i', 'ia', 'iao', 'ie', 'iu', 'ian', 'iang', 'in', 'ing', 'iong', 'u', 'ua', 'uo', 'ue', 'ui', 'uai', 'uan', 'uang', 'un', 'v', 've']

——————————————————

Numpy Array:
[ True  True  True  True  True  True  True  True  True  True  True False
  True False False False False False False False False False False False
 False False False False False False False False False False]
Displaying 1 of 24 rows within the array

Pandas DataFram

## Performing the Tests
To best compare the size and performance speed of each method this test works with two examples: a small two-letter phrase that is an acceptable pinyin term ("ba"), and a longer five-letter term that is not an acceptable pinyin term ("quang"). The %timeit function measures processing time with each method after running 100,000 iterations.

### Methods
1. The Python "in" function: the script searches through the list of acceptable pinyin terms for the supplied example.
2. The examples split into their initial and final pairs, their indexes derived from within the list of initials and finals, and those values are then used to find the appropriate boolean within the NumPy Array.
3. A Pandas DataFrame houses the boolean values, and three Pandas-specific methods are used to slice it:
    1. Traditional slicing method (to be deprecated in future versions): "exampleDF[initial][final]"
    2. The .loc method: "exampleDF.loc[initial][final]"
    3. The .iloc method: "exampleDF.iloc[index of initial within list of initials][index of final within list of finals]"

In [3]:
# Variables used for testing
smEval1 = 'ba'
smInitial1 = 'b'
smFinal1 = 'a'

smEval2 = 'quang'
smInitial2 = 'q'
smFinal2 = 'uang'


# Subject 1: Short, acceptable term
print('# Results of \"ba\" (short, valid):')

# Method 1: Python List
print('Python List: ' + str(sys.getsizeof(smList)))
%timeit - n 100000 smEval1 in smList

# Method 2: Numpy Array
print('\nNumpy array: ' + str(sys.getsizeof(initList) +
                              sys.getsizeof(finList) +
                              sys.getsizeof(smAr)) +
      ', without initials and finals lists: ' + str(sys.getsizeof(smAr)))
%timeit - n 100000 smAr[initList.index(smInitial1), finList.index(smFinal1)] != True

# Method 3: Pandas DataFrame
# 3a: slice
print('\nDataFrame: ' + str(sys.getsizeof(smDF)))
print('slice')
%timeit - n 100000 smDF[smFinal1][smInitial1] != True

# 3b: loc
print('loc')
%timeit - n 100000 smDF.loc[smInitial1, smFinal1] != True

print('iloc')
# 3c: iloc
%timeit - n 100000 smDF.iloc[initList.index(smInitial1), finList.index(smFinal1)] != True


# Subject 2: Longer, unacceptable term
print('\n\n# Results of \"quang\" (longer, invalid):')

# Method 1: Python List
print('Python List: ' + str(sys.getsizeof(smList)))
%timeit - n 100000 smEval2 in smList

# Method 2: Numpy Array
print('\nNumpy array: ' + str(sys.getsizeof(initList) +
                              sys.getsizeof(finList) +
                              sys.getsizeof(smAr)) +
      ', without initials and finals lists: ' + str(sys.getsizeof(smAr)))
%timeit - n 100000 smAr[initList.index(smInitial2), finList.index(smFinal2)] != True

# Method 3: Pandas DataFrame
# 3a: slice
print('\nDataFrame: ' + str(sys.getsizeof(smDF)))
print('slice')
%timeit - n 100000 smDF[smFinal2][smInitial2] != True

# 3b: loc
print('loc')
%timeit - n 100000 smDF.loc[smInitial2, smFinal2] != True

print('iloc')
# 3c: iloc
%timeit - n 100000 smDF.iloc[initList.index(smInitial2), finList.index(smFinal2)] != True

# Results of "ba" (short, valid):
Python List: 3760
399 ns ± 11.9 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

Numpy array: 1672, without initials and finals lists: 928
1.23 µs ± 7.34 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

DataFrame: 2427
slice
8.57 µs ± 39.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
loc
6.99 µs ± 51.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
iloc
8.52 µs ± 97 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


# Results of "quang" (longer, invalid):
Python List: 3760
3.68 µs ± 22.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

Numpy array: 1672, without initials and finals lists: 928
1.78 µs ± 34.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

DataFrame: 3067
slice
8.61 µs ± 49.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
loc
7.12 µs ± 37.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
iloc
9.2 µs ± 47.4 ns per loop (mean ± 

## Comparison: Size
The Pandas DataFrame would initially be the smallest form of a boolean table (2.427kb). However, it becomes 25% larger when reaccessed for multiple inquiries due to being loaded into its self-maintained cache (3.067kb). The NumPy array, in addition to the necessary initials and finals lists, takes more space than the initial size of the DataFrame but maintains a small size after repeated uses (2.6kb). The Python list, as expected, takes up the most amount of space due to its listing every possible syllable (3.76kb). Therefore, the NumPy array is the best option assuming the function evaluates multiple queries.
## Comparison: Speed
As expected, there is little difference in speed between the NumPy array evaluating the smaller, acceptable syllable and the larger, unacceptable syllable, with only a 0.55-microsecond difference between them. The Pandas DataFrame performed just as consistently, with an average of a 0.28-microsecond difference between the smaller and larger syllables. However, the DataFrame speed as a whole was impractically slower, with an average of 8.17 microseconds per 100,000 iterations, compared to 1.51 microseconds for the NumPy array.

The Python list resulted in the most significant disparity in time with a 3.28-microsecond difference between the smaller and larger syllables. However, the incredibly quick .399 microsecond result for the smaller, acceptable phrase initially made the option the most attractive for use. Having to search an entire list when supplied an invalid syllable is the disadvantage, leading to its 3.68-microsecond result. Future calculations finding the average length of syllables in practical Chinese use could dictate the real average time of the Python list option, but—assuming that the statistical mean suffices for optimizing this function—the 2.04-microsecond average does not beat that of the NumPy array. Therefore, the array is the best method for this function.

# The sylCount function
With the NumPy array chosen as the index method for acceptable vs. unacceptable syllables, the sylCount function evaluates the inputted text character-by-character. Computational processes benefit from the consistent syllable structure of the Chinese language because—with very few exceptions—most syllables begin with consonants as their initial and vowels as their final (e.g. "chou:" "ch" as initial, "ou" as final). Therefore, an automated process can quickly establish the initial until it encounters a vowel, accessing the array afterward to determine whether the combination of initial and final is valid.
![Example of function process](img/EvaluationExample.pdf)

The validity of syllables is an essential step in the function because not only will it determine whether the current sequence of letters counts as a syllable, but also if the syllable is still valid with additional letters added to it. For example, "tiao." The character-by-character loop within the function sets the initial to "t," and then on its next iteration identifies "i" as the final. "Ti," in itself, is a valid syllable. However, the following character "a" results in "tia," which is not a valid syllable. Searching the array is crucial at this stage because "tiao" is an acceptable syllable, but only if detected during the "tia" evaluation. If no other possibilities exist when evaluating a sequence, then it is safe to assume the current syllable has ended.

There are exceptions to this rule, specifically with the possibility of consonants appearing in the final, e.g., "qing." In the pinyin standard, "n," "ng," and "r" (only after following an "e") are all possible consonants in the final. The function, accordingly, must include the ability to evaluate these occurrences and assess whether they are finals, initials, or a combination of both in the case of "ng." Thankfully, these occurrences can be broken down into consistent patterns that are easy to translate into Python.

## The $n$ logic
The tables below illustrate the logic behind the evaluation of $n$. Because $n$ can appear as initial or final, the first and most straightforward approach to identifying its pattern is to look at each possible scenario for any character in pinyin. One character can have nine total possibilities in its placement, depending on whether there is a vowel, consonant, or nothing preceding it, and a vowel, consonant or nothing succeeding it. According to the table, $n$ is only an exception when there is a vowel preceding or succeeding it. All other possibilities are grammatically erroneous, especially if there is a preceding consonant.

In [28]:
from IPython.display import HTML, display
display(HTML("<table style='background-color:black;'><tr><td><img src='img/pinyinPairs.005.png'></td><td><img src='img/pinyinPairs.006.png'></td></tr></table>"))

These tables also indicate how the function processes valid possibilities. In the case of *n* with a preceding and succeeding vowel, the syllable count increments by one to account for it being the start of the next syllable. Identifying the character as initial or final is crucial for the function to evaluate the next character correctly.

## The *ng* logic

In [41]:
from IPython.display import HTML, display
display(HTML("<table style='background-color:black;'><tr><td><img src='img/pinyinPairs.003.png'></td><td><img src='img/pinyinPairs.004.png'></td></tr></table>"))

For *ng*, the logic is more explicit. The function only identifies the ng pair if there is a preceding vowel since all other possibilities are grammatically incorrect. Also, because *g* is assumed to be the current character in evaluation, the function can quickly identify whether it is the initial of a new syllable or the end of the final based on whether a vowel succeeds it.
## The *er* logic

In [42]:
from IPython.display import HTML, display
display(HTML("<table style='background-color:black;'><tr><td><img src='img/pinyinPairs.001.png'></td><td><img src='img/pinyinPairs.002.png'></td></tr></table>"))

The *er* pair has much greater complexity because all nine possibilities are grammatically correct. However, patterns emerge when analyzing their usage in pinyin. If er precedes any character, then the syllable count increments by one to account for the additional syllable. The count also increments by one if er—preceded by a vowel—is succeeded by anything except a vowel. Identifying the current character *r* as initial or final is a simple matter of evaluating whether the succeeding character is a vowel or not.

In [2]:
# Import necessary packages for function
import numpy as np
import csv

## REFERENCE DATA ##
# List of pinyin initials, finals for Numpy evaluation
PY = np.genfromtxt('data/pinyinDF.csv', delimiter=',', dtype=str)
initPYList = list(PY[1:, 0])
finPYList = list(PY[0, 1:])

# List of Wade-Giles initials, finals
WG = np.genfromtxt('data/wadegilesDF.csv', delimiter=',', dtype=str)
initWGList = list(WG[1:, 0])
finWGList = list(WG[0, 1:])

# Import Numpy arrays from CSV
arPY = PY[1:, 1:] == '1'
arWG = WG[1:, 1:] == '1'

## MAIN FUNCTION ##
# sylCount definition with parameters:
# crumbs: show breadcrumbs of NumPy evaluation
# errorReport: return error messages as strings, NaN for non-errors
# CBDB: detect duplication of family name within given name
# wg: evaluate in Wade-Giles standard


def sylCount(text, crumbs=False, errorReport=False, CBDB=False, wg=False):
    '''Estimate the number of Chinese characters based on inputted romanized text
    (amongst other things).'''

    ## REFERENCE VARIABLES ##
    # Lists/array for reference (pinyin set as default)
    vowel = ['a', 'e', 'i', 'o',
             'u', 'ü', 'v']  # for detecting switch from initial to final
    initList = initPYList  # list of initials
    finList = finPYList  # list of finals
    ar = arPY  # array of boolean values

    ## DATA EVALUATION ##
    # Massage data for evaluation
    # Replace ideographic symbols
    # Strip trailing/leading spaces
    # Convert to lowercase
    # If the process fails, the evaluated text is null
    replSym = {'—': '-',
               '–': '-',
               '\'': '’',
               '\u3000': ' '}
    try:
        for i, n in replSym.items():
            text = text.replace(i, n)
        text = text.strip().lower()
        evalText = text  # used for detection of invalid characters
        for n in replSym.values():
            evalText = evalText.replace(n, '').strip().lower()
    except:
        errorMsg = 'Error: null value'
        if crumbs:
            print(errorMsg)
        if errorReport:
            return errorMsg
        return [0]

    # If invalid characters are detected
    if not evalText.isalpha():
        errorMsg = 'Error: invalid non-alpha characters in name (' + \
            evalText + ')'
        if crumbs:
            print(errorMsg)
        if errorReport:
            return errorMsg
        return [0]

    # If hyphens or Wade-Giles are detected, change reference variables to WG:
    if '-' in text or wg:
        wg = True
        initList = initWGList
        finList = finWGList
        ar = arWG
        if crumbs:
            print('Switching to Wade-Giles: hyphen detected')

    # CBDB testing purposes ONLY: if the family name is duplicated in the given name
    if CBDB:
        dupTest = text.split()
        if len(dupTest) > 1:
            dupTestLen = [len(dupTest[0]), len(dupTest[1])]
            if (dupTest[0] == dupTest[1][0:dupTestLen[0]] and
                dupTestLen[1] > dupTestLen[0] and
                    dupTest[1][len(dupTest[0])] != '\''):
                errorMsg = 'Error: duplicate family name in given name'
                if crumbs:
                    print(errorMsg)
                if errorReport:
                    return errorMsg
                return [0]

    ## MAIN EVALUATION VARIABLES ##
    # Variables used for evaluation
    initial = ''
    final = ''
    newSyl = False
    syl = 1
    sylCount = []
    exFin = ''

    if crumbs:
        print('# Testing ' + text + ' #')

    ## MAIN EVALUATION ##
    # Cycle through each character in text
    for i, n in enumerate(text):

        ## SPECIAL CHARACTERS ##
        # Current character: Wade-Giles hyphen - new syllable
        if n == '-' and wg:
            step = '0. new syl'
            notes = 'hyphen'
            newSyl = True

        # Current character: space - new word
        elif n == ' ':
            step = '0. new word'

            # If double-spaced, skip to next character
            if text[i+1] == ' ':
                notes = 'discard extra space'

            # Otherwise, save syllable count to list and start next word
            else:
                sylCount.append(syl)
                newSyl = True
                notes = 'space'

        ## FINALS ##
        # If final has already been set, regardless of current character
        elif final:
            step = 'b. ' + initial + '|' + final + n

            # If current character is an apostrophe (as with Wade-Giles)
            if n == '’':
                newSyl = True
                notes = 'new syl - apostrophe'

            # If 'er,' 'ng,' or 'n'
            elif (final[-1] == 'e' and n == 'r' or
                  final[-1] == 'n'):
                notes = final[-1] + n

                # If there are more characters ahead, mark as new syllable
                if i != len(text) - 1 and text[i+1] != ' ':
                    newSyl = True
                    exFin = final[-1] + n
                    notes += ' - not \"' + exFin + '\" final'

                # Else if there were other vowels before 'er,' add a syllable
                elif (notes == 'er' and len(final) > 1) or n in vowel:
                    syl += 1
                    notes += ' - extra syl at end'

            # Check array with current initial + final combo
            else:
                notes = 'fin'

                # If there's an error, consider it as not a valid syllable
                try:
                    result = ar[
                        initList.index(initial),
                        finList.index(final + n)
                    ]
                    notes += (' - array lookup ' + initial + final + n +
                              ': ' + str(result))
                except:
                    result = False
                    notes += ' - array error'

                # If not a valid syllable, is it part of a longer syllable?
                if not result:
                    listFins = list(
                        i for i in finList if i.startswith(final+n))
                    if not listFins or not any(
                        i for i in listFins if ar[
                            initList.index(initial),
                            finList.index(i)
                        ]
                    ):    # It is!
                        newSyl = True
                        notes += ' - new syl'
                    else:  # It isn't.
                        final += n
                        notes += ' - definitely more'

                # Otherwise, the syllable is valid
                else:
                    final += n
                    notes += ' - maybe more'

        ## INITIALS ##
        # If there is no final set (then must be initial)
        else:
            step = 'a. ' + initial + n
            notes = 'init'

            # If character is not a vowel
            if n not in vowel:

                # Check if character is (or is part of) a valid initial
                if not any(i for i in initList if i.startswith(initial + n)):
                    errorMsg = ('Error: ' + initial + n +
                                ' - quitting, invalid initial')
                    if crumbs:
                        print(errorMsg)
                    if errorReport:
                        return errorMsg
                    return [0]
                initial += n
                notes += ' - cons'

            # Or if the vowel is the first character of the syllable
            elif not initial:
                initial = 'ø'
                final = n
                notes += ' - starting vowel'

            # Then the vowel is the first character of the final
            else:
                final = n
                notes += ' - vowel'

        ## DEBUG NOTES ##
        # Printed if crumbs parameter is true
        if crumbs:
            print(step + ': ' + notes)

        ## NEW SYLLABLE ##
        # Set up next initial and/or final based on prior evaluation above
        if newSyl:

            # If current letter is a vowel
            if n in vowel:

                # If final was being evaluated for 'n' (e.g. 'ch|iN-A|ng')
                if exFin:
                    initial = 'n'

                # Otherwise it's simply a new syllable (e.g. 'ch|eng-A|ng')
                else:
                    initial = 'ø'
                syl += 1
                final = n

            # Or it's anything but a space (consonant, hyphen, apostrophe)
            elif n != ' ':

                # If character is break symbol or next character is next syllable
                if (
                    (exFin and text[i+1] not in vowel) or
                    (n == '’' or n == '-')
                ):
                    initial = ''

                    # If it's the 'er' situation with a vowel before it
                    if exFin == 'er' and len(final) > 1:
                        syl += 1

                # Then current character is the start of next syllable
                else:
                    initial = n
                exFin = False
                final = ''
                syl += 1

            # Then it is a space (new word)
            else:
                initial = ''
                final = ''
                syl = 1
            if crumbs:
                print('new syl: ' + str(syl))
            newSyl = False

    # Append syllable count to list after end of text evaulation
    sylCount.append(syl)

    # Return a null value if the errorReport parameter is true
    if errorReport:
        return np.nan

    # Otherwise return the list of syllable counts
    else:
        return sylCount

## Case Study 1: Tang History Database
The Tang History Database, initiated by East Asian Studies Professor Anna Shields and designed by Data and Project Coordinator Jeff Heller, is a comparative database that allows scholars to visualize and analyze information from the Tang Dynastic History written by the Later Jin dynasty in 945AD and the version published by the Song dynasty in 1044AD. The project provides textual analysis tools such as dynamic search capabilities and a social network of Tang dynasty nobles and political figures. A challenge in preparing the data for this work was adequately tagging the names of figures within biographies. By doing this, we could better understand the relationship between Tang dynasty figures and how the text represents them, as well as the commonality of specific names used for a particular person, including instances where a person's biography mentions their name multiple times in various ways.

With Latin-based languages we separate a person's given name (first name) from their family name (last name) with the space character (" "), and from a computational standpoint this makes it easier for finding occurrences of the given name throughout the text (e.g., "This is **Jeff Heller's** biography. **Jeff** was born on..."). The Chinese language, however, does not provide a method of separation between names which is much harder to process computationally, especially if a name can also be a colloquial term (in English this occurs with names like "Summer" or "Shields"). Therefore, finding all occurrences of a person by their given name within their biography requires a customized method to identify it first. Harvard University's Chinese Biographical Database Project ([CBDB](https://projects.iq.harvard.edu/cbdb/home)) is an essential tool for this situation because it contains names in both the original Chinese and their pinyin counterpart.

In [44]:
# Import necessary packages
import pandas as pd

# Import name data to be split
names = pd.read_csv('data/TangHistoryNames.csv', index_col=0,
                    names=['EngName', 'ChiName'])

# 
names['ChiName'] = names['ChiName'].str.replace('\u3000', ' ').replace(' ', '')

names['syl'] = names['EngName'].apply(sylCount)


def nameSplit(chiName, syls, selector=0, collect=False):

    if len(syls) < 2:
        return None
    else:
        names = []

    if collect:
        for i, n in enumerate(syls):
            cut = sum(syls[:i])
            names.append(chiName[cut:cut+n])
    else:
        cut = sum(syls[:selector])
        names = chiName[cut:cut+syls[selector]]

    return names


names['names'] = names.apply(lambda x: nameSplit(
    x['ChiName'], x['syl'], collect=True), axis=1)
display(names)

EngName ChiName     syl       names
379873       An Lushan     安祿山  [1, 2]     [安, 祿山]
380462       An Qingxu     安慶緒  [1, 2]     [安, 慶緒]
378063       An Sishun     安思順  [1, 2]     [安, 思順]
378621      An Taiqing     安太清  [1, 2]     [安, 太清]
162825      An Xinggui     安興貴  [1, 2]     [安, 興貴]
378334       An Xiuren     安修仁  [1, 2]     [安, 修仁]
376419    Ashina Sheer   阿史那社爾  [3, 2]   [阿史那, 社爾]
377887  Ashina Sunishi  阿史那蘇尼失  [3, 3]  [阿史那, 蘇尼失]
384386    Ashina Zhong    阿史那忠  [3, 1]    [阿史那, 忠]
32227         Bai Juyi     白居易  [1, 2]     [白, 居易]
32225     Bai Minzhong     白敏中  [1, 2]     [白, 敏中]
167504          Bai Qi      柏耆  [1, 1]      [柏, 耆]
377724      Bai Xiaode     白孝德  [1, 2]     [白, 孝德]
32228     Bai Xingjian     白行簡  [1, 2]     [白, 行簡]
376198   Bai Yuanguang     白元光  [1, 2]     [白, 元光]
189643     Bai Zhizhen     白志貞  [1, 2]     [白, 志貞]
376103    Ban Jingqian     班景倩  [1, 2]     [班, 景倩]
377526        Bao Fang      鮑防  [1, 1]      [鮑, 防]
92411           Bao Ji      包佶  [1, 1]      [包, 佶]
18384           Bi Gou      畢構  [1, 1]      [畢, 構]
145777        Bi Shang      畢坰  [1, 1]      [畢, 坰]
18386          Bi Xian      畢諴  [1, 1]      [畢, 諴]
18385            Bi Xu      畢栩  [1, 1]      [畢, 栩]
380260        Cai Xide     蔡希德  [1, 2]     [蔡, 希德]
170542         Cao Hua      曹華  [1, 1]      [曹, 華]
184606         Cao Que      曹確  [1, 1]      [曹, 確]
31628       Cen Wenben     岑文本  [1, 2]     [岑, 文本]
377557       Chai Shao      柴紹  [1, 1]      [柴, 紹]
376066       Chang Cui      暢璀  [1, 1]      [暢, 璀]
94490        Chang Gun      常衮  [1, 1]      [常, 衮]
...                ...     ...     ...         ...
93891        Zheng Yao      鄭繇  [1, 1]      [鄭, 繇]
183952       Zheng Yin      鄭愔  [1, 1]      [鄭, 愔]
93853        Zheng Yin      鄭絪  [1, 1]      [鄭, 絪]
12766     Zheng Yin(2)      鄭絪     [0]        None
376122      Zheng Yuan      鄭元  [1, 1]      [鄭, 元]
93808    Zheng Yuanshu     鄭元璹  [1, 2]     [鄭, 元璹]
33159      Zheng Yunmu     鄭允謨  [1, 2]     [鄭, 允謨]
33153     Zheng Yuqing     鄭餘慶  [1, 2]     [鄭, 餘慶]
428985     Zheng Zhide     鄭祇徳  [1, 2]     [鄭, 祇徳]
189319       Zheng Zhu      鄭注  [1, 1]      [鄭, 注]
152251  Zhong Shaojing     鍾紹京  [1, 2]     [鍾, 紹京]
442513        Zhou Chi      周墀  [1, 1]      [周, 墀]
186479     Zhou Faming     周法明  [1, 2]     [周, 法明]
376385       Zhou Xing      周興  [1, 1]      [周, 興]
182055    Zhou Yunyuan     周允元  [1, 2]     [周, 允元]
377579   Zhou Zhiguang     周智光  [1, 2]     [周, 智光]
16604       Zhu Jingze     朱敬則  [1, 2]     [朱, 敬則]
92527           Zhu Pu      朱朴  [1, 1]      [朱, 朴]
94154      Zhu Qinming     祝欽明  [1, 2]     [祝, 欽明]
377600   Zhu Quanzhong     朱全忠  [1, 2]     [朱, 全忠]
16603       Zhu Rengui     朱仁軌  [1, 2]     [朱, 仁軌]
380420         Zhu Tao      朱滔  [1, 1]      [朱, 滔]
194753       Zhu Xicai     朱希彩  [1, 2]     [朱, 希彩]
377948  Zhu Zhongliang     朱忠亮  [1, 2]     [朱, 忠亮]
194619          Zhu Zi      朱泚  [1, 1]      [朱, 泚]
380194     Zhuge Dewei    諸葛德威  [2, 2]    [諸葛, 德威]
93924       Zong Chuke     宗楚客  [1, 2]     [宗, 楚客]
94145        Zu Junyan     祖君彥  [1, 2]     [祖, 君彥]
92290      Zuo Nandang     左難當  [1, 2]     [左, 難當]
381201     Zuo Youxian     左遊仙  [1, 2]     [左, 遊仙]

[1609 rows x 4 columns]

In [88]:
import pandas as pd

names = pd.read_csv('data/allCBDBnames.csv', index_col=0,
                    names=['EngName', 'ChiName'], header=0)

replSym = {'\'': '’',
           '\u3000': ' '}

names['ChiName'] = names['ChiName'].apply(lambda x: x.str.replace())

names['ChiName'] = names['ChiName'].str.replace(' ', '')

names['syl'] = names['EngName'].apply(sylCount)

names['errors'] = names['EngName'].apply(sylCount, CBDB=True,
                                         errorReport=True).fillna('')


def tryAdd(x):
    try:
        return sum(x)
    except:
        return


names['check'] = names['ChiName'].str.len() == names['syl'].apply(tryAdd)

names[names.errors.str.contains('duplicate')][names.check == False][~names.ChiName.str.contains(
    '\(')].to_csv('CBDB_familyNameDupes.csv')

/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:20: UserWarning: Boolean Series key will be reindexed to match DataFrame index.


In [27]:
import pandas as pd

names = pd.read_csv('data/allCBDBnames.csv', index_col=0,
                    names=['EngName', 'ChiName'], header=0)

names = names.sample(n=10)
names['syl'] = names['EngName'].map(sylCount)
names.iloc[0]['syl'][1]

2

In [34]:
testEng = 'changwangzhong liang sheertiao'
testChi = '莊欽祖莊有信欽'
testResult = sylCount(test)
print(testResult)

def nameSplit(chiName, syls, selector=0, collect=False):

    if len(syls) < 2:
        return None
    else:
        names = []

    if collect:
        for i, n in enumerate(syls):
            cut = sum(syls[:i])
            names.append(chiName[cut:cut+n])
    else:
        cut = sum(syls[:selector])
        names = chiName[cut:cut+syls[selector]]

    return names


print(nameSplit(testChi, testResult, collect=True))

[3, 1, 3]
['莊欽祖', '莊', '有信欽']
